File: src/nfan/fan/sys/Weekday.cs
//
// Copyright (c) 2007, Brian Frank and Andy Frank
// Licensed under the Academic Free License version 3.0
//
// History:
//   23 Feb 07  Andy Frank  Creation
//

namespace Fan.Sys
{
  /// <summary>
  /// Weekday.
  /// </summary>
  public sealed class Weekday : Enum
  {

    public static readonly Weekday m_sun = new Weekday(0, "sun");
    public static readonly Weekday m_mon = new Weekday(1, "mon");
    public static readonly Weekday m_tue = new Weekday(2, "tue");
    public static readonly Weekday m_wed = new Weekday(3, "wed");
    public static readonly Weekday m_thu = new Weekday(4, "thu");
    public static readonly Weekday m_fri = new Weekday(5, "fri");
    public static readonly Weekday m_sat = new Weekday(6, "sat");

    internal static readonly Weekday[] array =
    {
      m_sun, m_mon, m_tue, m_wed, m_thu, m_fri, m_sat
    };

    public static readonly List m_values = new List(Sys.WeekdayType, array).ro();

    private Weekday(int ordinal, string name)
    {
      Enum.make_(this, Int.m_pos[ordinal], Str.make(name).intern());
      this.ord = ordinal;
      this.localeAbbrKey  = Str.make(name + "Abbr");
      this.localeFullKey  = Str.make(name + "Full");
    }

    public static Weekday fromStr(Str name) { return fromStr(name, Bool.True); }
    public static Weekday fromStr(Str name, Bool check)
    {
      return (Weekday)doFromStr(Sys.WeekdayType, name, check);
    }

    public override Type type() { return Sys.WeekdayType; }

    public Weekday increment() { return array[(ord+1)%array.Length]; }

    public Weekday decrement() { return ord == 0 ? array[array.Length-1] : array[ord-1]; }

    public Str toLocale() { return toLocale(null); }
    public Str toLocale(Str pattern)
    {
      if (pattern == null) return localeAbbr();
      if (pattern.isEveryChar('W'))
      {
        switch (pattern.val.Length)
        {
          case 3: return localeAbbr();
          case 4: return localeFull();
        }
      }
      throw ArgErr.make("Invalid pattern: " + pattern).val;
    }

    public Str localeAbbr() { return abbr(Locale.current()); }
    public Str abbr(Locale locale)
    {
      return locale.get(Str.sysStr, localeAbbrKey);
    }

    public Str localeFull() { return full(Locale.current()); }
    public Str full(Locale locale)
    {
      return locale.get(Str.sysStr, localeFullKey);
    }

    public static Weekday localeStartOfWeek()
    {
      return fromStr(Locale.current().get(Str.sysStr, localeStartKey));
    }

    static readonly Str localeStartKey = Str.make("weekdayStart");

    internal readonly int ord;
    readonly Str localeAbbrKey;
    readonly Str localeFullKey;
  }
}

File: src/jfan/fan/sys/Weekday.java
//
// Copyright (c) 2006, Brian Frank and Andy Frank
// Licensed under the Academic Free License version 3.0
//
// History:
//   19 Sep 06  Brian Frank  Creation
//
package fan.sys;

/**
 * Weekday
 */
public final class Weekday
  extends Enum
{

  public static final Weekday sun = new Weekday(0, "sun");
  public static final Weekday mon = new Weekday(1, "mon");
  public static final Weekday tue = new Weekday(2, "tue");
  public static final Weekday wed = new Weekday(3, "wed");
  public static final Weekday thu = new Weekday(4, "thu");
  public static final Weekday fri = new Weekday(5, "fri");
  public static final Weekday sat = new Weekday(6, "sat");

  static final Weekday[] array = { sun, mon, tue, wed, thu, fri, sat };

  public static final List values = new List(Sys.WeekdayType, array).ro();

  private Weekday(int ordinal, String name)
  {
    Enum.make$(this, FanInt.pos[ordinal], name.intern());
    this.ord = ordinal;
    this.localeAbbrKey  = name + "Abbr";
    this.localeFullKey  = name + "Full";
  }

  public static Weekday fromStr(String name) { return fromStr(name, true); }
  public static Weekday fromStr(String name, boolean checked)
  {
    return (Weekday)doFromStr(Sys.WeekdayType, name, checked);
  }

  public Type type() { return Sys.WeekdayType; }

  public Weekday increment() { return array[(ord+1)%array.length]; }

  public Weekday decrement() { return ord == 0 ? array[array.length-1] : array[ord-1]; }

  public String toLocale() { return toLocale(null); }
  public String toLocale(String pattern)
  {
    if (pattern == null) return localeAbbr();
    if (FanStr.isEveryChar(pattern, 'W'))
    {
      switch (pattern.length())
      {
        case 3: return localeAbbr();
        case 4: return localeFull();
      }
    }
    throw ArgErr.make("Invalid pattern: " + pattern).val;
  }

  public String localeAbbr() { return abbr(Locale.current()); }
  public String abbr(Locale locale)
  {
    return locale.get("sys", localeAbbrKey);
  }

  public String localeFull() { return full(Locale.current()); }
  public String full(Locale locale)
  {
    return locale.get("sys", localeFullKey);
  }

  public static Weekday localeStartOfWeek()
  {
    return fromStr(Locale.current().get("sys", "weekdayStart"));
  }

  final int ord;
  final String localeAbbrKey;
  final String localeFullKey;
}
